### ISSUE TITLE
Frequent lockups and calendar entry disappearance

### ISSUE THREAD

**Issue by fxshpac** [mpcjanssen/simpletask-android#744]
I'm currently on latest 9.0.2 of ST Cloudless. Since recently I have been experiencing frequent loss of response from the app while performing regular operations such as editing individual entries, scroll them up and down, etc. I have about 90 tasks in the app and 15 out them have due/threshold dates set. The Calendar Reminder feature is also enabled. In other occasions I found all the calendar entries were gone and open/exit the app most of time can bring them back. Not sure if these two issue are related though. Anyway, let me know what data I can collect to help troubleshooting these issues.

**Thanks**

**Comment by mpcjanssen** [comment]
First question, do you use something like a battery saver or background app killer?

Secondly can you send me a log from settings other debug log?

**Comment by fxshpac** [comment]
Thanks @mpcjanssen for getting back to me. I noticed a much better outcome after I made sure power saving mode is not enabled for this particular app. Before it was set to Automatic mode. However, I can still see brief UI lockups but mostly it's right after I changed due/threshold dates for certain tasks. I believe the app was trying to refresh the Calendar entries at that time because if I go to Calendar app right after that I can see all ST Cloudless entries are being removed and then added back. That's pretty heavy operation. I wonder it can be further optimized by only updating the ones which were changed. And it should eliminate the UI lockup for sure. Just a thought.

For testing purpose, I also disabled Calendar Reminder feature and that yielded the best result where no lockup ever happened during the testing period. So that further tells me the issue has something to do with the feature implementation.

With that I can still send you the debug log if you would like to look at it.

**Comment by vojtechkral** [comment]
>  I believe the app was trying to refresh the Calendar entries at that time because if I go to Calendar app right after that I can see all ST Cloudless entries are being removed and then added back. That's pretty heavy operation. I wonder it can be further optimized by only updating the ones which were changed. 

Yes, it definitely can! I even have some code that does that. I have just been procrastinating on it / preoccupied with other stuff and didn't bring myself to finalize it. I really should get on that.

**Comment by mpcjanssen** [comment]
What i find a bit surprising in all this, is that the calendar sync happens in the background, so i wouldn't expect it to have such an impact.

**Comment by vojtechkral** [comment]
FYI I have a working implementation of a diffing notifications sync, it seems to work, but I need to do some more testing and code cleanup. Will report back...

**Comment by mpcjanssen** [comment]
@fxshpac any update since the latest versions which include the optimized calendar sync algorithm?

**Comment by smichel17** [comment]
Is calendar sync happening on the UI thread? (I can check myself later today when not on mobile)

**Comment by vojtechkral** [comment]
> Is calendar sync happening on the UI thread?

No, and it ever was that way.

**Comment by fxshpac** [comment]
@mpcjanssen Just did a quick test with version 9.0.6 the latest from Play store. Currently I have 22 entries set with Due dates. When I change due date of an individual entry in UI and continue looking around in UI I don't experience the sudden lockups as experienced before. That's a good progress to me. One thing to keep in mind is that I upgrade my phone from Note 4 to S8+ so I'm not sure if that helps in certain way.

After I leave the app to check my Calendar, I notice there is a noticeable delay to see the entries there to be updated. Based on #752, this is expected behavior due to 20s delay timer. Wonder if the timer value can be configurable though.

That's all I can see for now. Thank you and @vojtechkral for the effect to make the app even better.

**Comment by mpcjanssen** [comment]
@fxshpac thanks for confirming it improved the situation.
The delay is part of the reason why it performs better (the diff algorithm is another part) I personally don't really see it as an issue that the initial creation takes a bit of time and don't really see the value in making it configurable (with all possible side effects). So unless there is a really pressing reason to make it configurable it will stay like this.
